### PR TITLE
corrfix

### PR DESCRIPTION
**Pull request by knaaptime**
- add first draft of correlogram
- add comment to include geoda version
- add correlogram test
- typos, test func
- add explanation and some notebook text
- update to pyproject
- scm init
- update release workflow
- use radius
- CI: properly test min and dev (#258)
- rebase
- rm manifest

### CODE REVIEW

**Comment by knaaptime** [comment]
supercedes #253

**Comment by knaaptime** [comment]
there's nothing harder than pulling off a successful rebase. I'll die on this hill.

**Comment by jGaboardi** [comment]
> there's nothing harder than pulling off a successful rebase. I'll die on this hill.

File: esda/correlogram.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        geodataframe holding spatial and attribute data
+    variable: str
+        column on the geodataframe used to compute autocorrelation statistic
+    distances : list
+        list of distances to compute the autocorrelation statistic
+    distance_type : str, optional
+        which concept of distance to increment. Options are {`band`, `knn`}.
+        by default 'band' (for `libpysal.weights.DistanceBand` weights)
+    statistic : str, by default 'I'
+        which spatial autocorrelation statistic to compute. Options in {`I`, `G`, `C`}
```
Comment (martinfleis) [inline esda/correlogram.py:76]: could you add a key to what I, G, C mean? Saying that it is Moran's I, Geary C, Getis-Ord G?

File: esda/correlogram.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        if max(distances) > gdf.shape[0] - 1:
+            with ValueError as e:
+                raise e("max number of neighbors must be less than or equal to n-1")
+        W = KNN
+    else:
+        with NotImplementedError as e:
+            raise e("distance_type must be either `band` or `knn` ")
+
+    #  should be able to build the tree once and reuse it?
+    #  but in practice, im not seeing any real difference from starting a new W from scratch each time
```
Comment (martinfleis) [inline esda/correlogram.py:119]: Not even for large data? I suppose that repeated creation of the tree could make a dent on this.

File: esda/tests/test_correlogram.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import geopandas as gpd
+
+
+sac = gpd.read_file(examples.load_example("Sacramento1").get_path("sacramentot2.shp"))
+sac = sac.to_crs(sac.estimate_utm_crs())  # now in meters)
+
+distances = [i + 500 for i in range(0, 2000, 500)]
+
+def test_correlogram():
```
Comment (martinfleis) [inline esda/tests/test_correlogram.py:14]: Do we have some ecosystem-wide standards on what is the minimal acceptable test suite?

Comment (ljwolf) [reply esda/tests/test_correlogram.py:14]: No... We've talked about separating "correctness" tests (does the statistic recover the same result every time?) and option tests/user tests (does the function handle all combination of arguments/input types correctly?), but we've never agreed on what is necessary for any contribution...

File: esda/correlogram.py (hunk 2 above)
Comment (knaaptime) [reply esda/correlogram.py:119]: I was "testing" on a dataset with ~65k points, but just not terribly thoroughly

what i meant was, I'd also done `KNN.from_dataframe(df, k=k)` in each loop, instead of `W(tree)`, where the former needs to build the tree each time in addition to querying it. Then i figured it might be faster to re-use the same tree repeatedly instead

what i was hoping for was something like @ljwolf said, where i'd do the expensive call once for the largest distance, then repeatedly use the same structure for repeated calls (sorta like how pandana works with `preprocess`). If we just loop over k sorted in descending order, does that help?

File: esda/correlogram.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        stat_kwargs = dict()
+    if weights_kwargs is None:
+        weights_kwargs = dict()
+    if statistic == "I":
+        STATISTIC = Moran
+    elif statistic == "G":
+        STATISTIC = G
+    elif statistic == "C":
+        STATISTIC = Geary
+    else:
```
[the remark this answers is not in the document]
Comment (knaaptime) [reply esda/correlogram.py:103]: cool, that makes sense. I think the first version I wrote actually took the Moran class as the first argument, but i thought the dispatching actually helped clarify what the function was doing. The internals ended up a lot more generic than i figured they'd need to be

so the most abstract version returns the output of a callable for a range of W/Graph specifications (in table form), and there are probably plenty of uses for that. But then do we want to call the function something more generic since it no longer necessarily computes some measure of correlation?

File: esda/correlogram.py (hunk 2 above)
Comment (ljwolf) [reply esda/correlogram.py:119]: (a) I think it's fine to keep this current implementation for now. 
(b) yeah basically: a more performant implementation would probably use `max(distances)` to compute `.sparse_distance_matrix()` (or built a sparse matrix from the output of `query()` if KNN) one time, and then progressively mask that based on each smaller distance value... basically, we want to query once, rather than repeat the query to yield results we can "precompute".

File: esda/correlogram.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ljwolf) [reply esda/correlogram.py:103]: Totally, and then define a specific `correlogram()` function that only supports Moran, spatial Pearson, or spatial tau? 

For the general function, @TaylorOshan and I have been calling these "profile" functions ([bottom of page 302](https://github.com/pysal/esda/files/12262367/s10109-022-00384-8.pdf)), so I'm partial to `distance_profile()` or `spatial_profile()`.

[the remark this answers is not in the document]
Comment (ljwolf) [reply esda/correlogram.py:103]: Well, I guess even like.... I have a nascent implementation of a nonparametric robust Moran/Local Moran statistic that would benefit from being able to plug in directly? I think it's OK to call the function `correlogram()`, and then allow for a user-defined callable. The fact that it's user-defined means that we can't enforce the output is callable, and this keeps things more simple for the user.

[the remark this answers is not in the document]
Comment (knaaptime) [reply esda/correlogram.py:103]: profile is the same nomenclature [over in segregation](https://pysal.org/segregation/api.html#dynamics) but maybe we should centralize some of this logic then

File: esda/correlogram.py (hunk 2 above)
Comment (knaaptime) [reply esda/correlogram.py:119]: gotcha. [Same way](https://github.com/oturns/geosnap/blob/main/geosnap/analyze/network.py) i did it for the isochrones with pandana. Get the adjacency once and trim it down for each distance. So here it would be something like 

```python

maxtree = KDTree(max(distances))
mintree = KDTree(min(distances))

distmat = maxtree.sparse_distance_matrix(mintree)
for dist in distances
	w = Graph.build_knn(distmat, k=k)
```
or something?

Comment (knaaptime) [reply esda/correlogram.py:119]: i guess what im asking is, what's the most efficient way to do this with the new Graph? 

If we have a Graph based on the max distance, then it's already got all the information we need for every lesser distance, and all we need to do is filter on the adjlist where wij<distance. Presumably thats what simething like `Graph.build_knn(distmat, k=k)` would do when passed an adjlist or a sparse matrix because all it needs to do is subset the existing data, not rebuild the distance relationships

for this problem its easy enough to just build the adjlist once and literally filter it down each distance and instantiate a new Graph/W from that adjlist, but is that the best way?

Comment (martinfleis) [reply esda/correlogram.py:119]: > for this problem its easy enough to just build the adjlist once and literally filter it down each distance and instantiate a new Graph/W from that adjlist, but is that the best way?

I suppose this would need to be tested. My sense is that a filter of adjacency based on a distance will be significantly slower than a tree query but I might be wrong.

File: esda/tests/test_correlogram.py (hunk 3 above)
Comment (martinfleis) [reply esda/tests/test_correlogram.py:14]: Yup, makes sense. I asked because I am not sure when should I be fine with the state of tests when doing the review, so an agreed guideline would be nice.

Comment (jGaboardi) [reply esda/tests/test_correlogram.py:14]: Very much agree with Martin on this. Seems like our general rule of thumb has been that somewhere between 70-90% is the acceptable coverage. I always try to get as close to 100% as possible to (attempt to) hit any weird edge cases, but I know that isn't required a lot of the time. If we want to decide on an actual number, this is surely something for at least the steering committee to vote on. Moreover, (1) as a step towards catching edge cases we may want to consider seeing about implementing [`hypothesis`](https://hypothesis.readthedocs.io/en/latest/), and (2) we should improve testing across all submodules to properly test minimum requirements, [as Martin points out here](https://github.com/pysal/esda/pull/258).

Comment (knaaptime) [reply esda/tests/test_correlogram.py:14]: >If we want to decide on an actual number, this is surely something for at least the steering committee to vote on. Moreover, (1) as a step towards catching edge cases we may want to consider seeing about implementing [hypothesis](https://hypothesis.readthedocs.io/en/latest/), and (2) we should improve testing across all submodules to properly test minimum requirements, https://github.com/pysal/esda/pull/258.

again, i dont think an ecosystem-wide policy is realistic, but definitely something we should discuss. The question is about cost-benefit. If we're too strict about test coverage, it will discourage contributions, and i think for many packages, that's a greater risk than having some untested code. I don't have time to get absorbed in edge-cases, so I'd prefer to let those issues surface before letting them block new functionality getting merged.

case in point, please feel free to write out the rest of the test suite here :)

Comment (jGaboardi) [reply esda/tests/test_correlogram.py:14]: > > If we want to decide on an actual number, this is surely something for at least the steering committee to vote on. Moreover, (1) as a step towards catching edge cases we may want to consider seeing about implementing [hypothesis](https://hypothesis.readthedocs.io/en/latest/), and (2) we should improve testing across all submodules to properly test minimum requirements, #258.
> 
> again, i dont think an ecosystem-wide policy is realistic, but definitely something we should discuss. The question is about cost-benefit. If we're too strict about test coverage, it will discourage contributions, and i think for many packages, that's a greater risk than having some untested code. I don't have time to get absorbed in edge-cases, so I'd prefer to let those issues surface before letting them block new functionality getting merged.
> 
> case in point, please feel free to write out the rest of the test suite here :)

FWIW @knaaptime I am "on your side" here (though I think we're all on the same side). In my opinion, getting new functionality merged should take precedence over a strict coverage number for individual PRs, so long as that new functionality is being "mostly" covered (where "mostly" can either be more lines covered *or* majority cases covered). More testing can be added later as time & energy permit. I hope my comments did not come off as combative.

Comment (knaaptime) [reply esda/tests/test_correlogram.py:14]: 😅 sorry my style is always blunt! no offense taken or intended. We're all on the same page. We need some guidelines on merging--definitely. I just want to avoid a situation where we let stuff sit here, because it's usually easier for us core maintainers to take incremental passes at improving test coverage, instead of imposing a big burden on the PR opener

(for this one, I certainly don't *need* this PR merged, but while it's sitting here it makes for a good example. I already have the function myself, but if other folks want to use it, i'd prefer not to let corner cases prevent that, since i know i don't have a ton of time to write out the test suite. I can usually commit to responding to bug reports though)

File: esda/correlogram.py (hunk 1 above)
Comment (jGaboardi) [reply esda/correlogram.py:76]: This is a nice suggestion.

File: esda/tests/test_correlogram.py (hunk 3 above)
Comment (martinfleis) [reply esda/tests/test_correlogram.py:14]: My question was primarily coming from my GeoPandas experience and understanding that what we aim for there is not applicable in PySAL. So I was just wondering what is the level people feel is enough. Totally agree with all you said above, there's no need to block stuff just because tests are not 100%. Happy to have a further chat about it during the next call. And happy to merge this with the test suite as is :).

Comment (jGaboardi) [reply esda/tests/test_correlogram.py:14]: Once again I agree with Martin, but maybe after addressing https://github.com/pysal/esda/pull/259#discussion_r1284119611?

File: esda/correlogram.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        for dist in distances
+    ]
+
+    outputs = Parallel(n_jobs=n_jobs, backend=backend)(
+        delayed(_get_autocorrelation_stat)(i) for i in inputs
+    )
+
+    return (
+        pd.DataFrame(outputs)
+        .select_dtypes(["number"])
```
Comment (knaaptime) [inline esda/correlogram.py:145]: so, in the abstract version, this function becomes `spatial_profile(callable, y, gdf, **kwargs)`, where `callable` is any function that takes (y,w) as arguments.

in that case, should we get rid of this `.select_dtypes` here, and return everything the callable has? Also, since [this line](https://github.com/knaaptime/esda/blob/f2b952e54933d17fd9d88ad82838f3c45d40b06a/esda/correlogram.py#L41) requires a *class*, not a generic callable.

Comment (ljwolf) [reply esda/correlogram.py:145]: I think just keep it as correlogram. Allowing a user defined function already means we can't assume/ensure the output is "correlation", but it's a relatively advanced way to work with the function.

And yes, the output type of the callable has to be a namespace/namedtuple/object kind of thing, but I think that's an ok requirement

File: esda/correlogram.py (hunk 2 above)
Comment (knaaptime) [reply esda/correlogram.py:119]: so, i'm playing with this, and it would be straightforward to filter the adjlist, which is fast, since its just `adj[adj['weight']<=distance]`. 

...But once we subset the W, now we have to subset the original dataframe to align it, which is something we said we wouldnt do. Is there a good way to do this?

**Comment by knaaptime** [comment]
this should be just about good to go, but it's still failing for stuff like join counts because the class stashes all sorts of things as attriubutes (like the W and its adjacency list) that cant get serialized by joblib, and LOSH/Spatial_Pearson which have a different signature that needs w.sparse

File: esda/correlogram.py (hunk 2 above)
Comment (martinfleis) [reply esda/correlogram.py]: To keep the structure, you can just assign weight=0 to everything above the set distance threshold.

**Comment by knaaptime** [comment]
Smart. The issue I was runnning into was that w.to_adjlist sorts the neighbors, so once you convert and try to subset the W, it and df are no longer aligned, which is a mess

--

Elijah Knaap, Ph.D.
Senior Research Scientist & Associate Director
Center for Open Geographical Science
San Diego State University
knaaptime.com | @knaaptime

File: esda/correlogram.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        additional keyword arguments passed to the `esda` autocorrelation statistic class.
+        For example for faster results with no statistical inference, set the number
+        of permutations to zero with {permutations: 0}
+    select_numeric : bool
+        if True, only return numeric attributes from the original class. This is useful
+        e.g. to prevent lists inside a "cell" of a dataframe
+    n_jobs : int
+        number of jobs to pass to joblib. If -1 (default), all cores will be used
+
+
```
Comment (jGaboardi) [inline esda/correlogram.py:91]:
```suggestion
```

File: esda/correlogram.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    if distance_type == "band":
+        W = DistanceBand
+    elif distance_type == "knn":
+        if max(distances) > gdf.shape[0] - 1:
+            with ValueError as e:
+                raise e("max number of neighbors must be less than or equal to n-1")
+        W = KNN
+    else:
+        with NotImplementedError as e:
+            raise e("distance_type must be either `band` or `knn` ")
```
Comment (jGaboardi) [inline esda/correlogram.py:111]:
```suggestion
            raise e("distance_type must be either `band` or `knn`")
```

**Comment by knaaptime** [comment]
what other tests do folks think are necessary for this?

**Comment by ljwolf** [comment]
I think at least: 
1. verify that the documented association measures work as expected
2. verify that a simple callable completes its computation

Then I think it's done!

**Comment by knaaptime** [comment]
cool. I think 1 is covered by the existing test using I? just need to add a generic callable test for 2?

**Comment by ljwolf** [comment]
> covered by the test using I

Ah ok, in light of the previous discussion, fine! I meant that we should test all of the estimators we specify to ensure correctness. 

Then, I think the last bit is just the test failures?

**Comment by ljwolf** [comment]
I'm trying to start a re-run of these tests, but when I think you've protected your remote @knaaptime? I made @jGaboardi's whitespace edit, and tried the following:

Screenshot 2024-08-16 at 12 02 20

**Comment by knaaptime** [comment]
i think there's something weird with my github at the moment. All my review comments say 'pending' too. Let me see

**Comment by knaaptime** [comment]
Screenshot 2024-08-16 at 8 29 09 AM
 just checked this guy. try again?

**Comment by martinfleis** [comment]
> Looking into the W vs. Graph choice here, one challenge moving this to the Graph object is that the W allowed users to pass a pre-constructed KDTree and simply request new W from this prebuilt tree, but Graph does not. This makes the correlogram() function performant. If we adjust Graph to allow build_distance_band/build_knn to accept a KDTree input, then a migration path is clearer.
> 
> For now, I can build the slower wrapper, but I think this should be discussed.

Okay, let's keep it as is and make a required change in Graph to allow this.

**Comment by knaaptime** [comment]
this all makes sense to me. Thanks @ljwolf !

**Comment by knaaptime** [comment]
@ljwolf thanks for the lowess version, looks great. Should we also implement/document that this version works a little differently [than geoda](https://geodacenter.github.io/workbook/5a_global_auto/lab5a.html#creating-a-spatial-correlogram), which, apart from using lowesss also seems to work on an expanding donut (i.e. it trims lower-bound distances as it moves outward?)

**Comment by ljwolf** [comment]
To explain more fully: 

the `_lowess_correlogram()` part estimates a lowess regression on the correlation and distance: 

`z_i*z_j = f(d_{ij}) u_{ij}`

This is the equation listed in the GeoDa documentation. Unfortuately, we have to parameterise this a little differently from GeoDa no matter what we do, unless we want to implement our own `lowess` regressor or simply use the average (unregularized) correlation in bin.

`statsmodels` lowess lets us specify the points at which the lowess is calculated (`xvals`) and *what fraction of the data should be used around each point* (`frac`). This second part is because their approach uses a (novel to me!) [interval knn-based method](https://github.com/statsmodels/statsmodels/blob/main/statsmodels/nonparametric/_smoothers_lowess.pyx#L267) to track points near `xvals`. 

Hence, we cannot quite do lowess over bins exactly like GeoDa; the bins must have dynamic width (in terms of distance) and represent a fixed fraction of the data. This has tradeoffs. Hence, I'm suggesting we deviate from GeoDa's parameterization and make the following tradeoffs: 

1. **set the end of the support to half the longest bounding box diagonal by default**. This allows us to ensure coverage for both parametric and nonparametric calculations without requiring us to build the full distance matrix up front. But, it's often longer than we want to visualize, since correlations often quickly go to zero. 
2. **use only the upper-triangle** of the covariance and distance matrices if the distance matrix is symmetric. GeoDa docs suggest using the full matrix is not always the best choice, and that they use it to avoid having bins with too-few observations. Given `statsmodels` lowess uses a `knn` local regression, we don't need to worry about having too few observations at each `xval` and the duplication just slows the routine down. If the distance matrix is not symmetric (as might happen if `metric='precomputed'`), we use the full matrix anyway.  
3. **set the lowess `frac` according to the typical fraction of data in each bin**. Here, note that `support` won't always span all distances. This minimizes the re-use of points (with no reuse when bins are constant). 
    (a) calculate the implied bin widths from user-supplied `support`, 
    (b) calculate the width of the first and last bin.
    (c) look up/down 1/2 the first/last bin width from the lowest/highest support value (bounded by zero on the left), and 
    (d) calculate the fraction of distances that fall within this span: `frac_span`, the fraction of point pairs spanned by bins implied by the support values. 
    (e) Then, we set the lowess `frac` to `frac_span/n_bins`. 

For evenly spaced bins, this will assign each observation to its nearest support, even if `support.max()+1/2*bin_width` leaves off some pairs. For unevenly-spaced support, this will recycle observations across bins. I don't think we can get around this with statsmodels' lowess.

**Comment by ljwolf** [comment]
This now has full tests and nonparametric is in the notebook. If we're fine with the defaults, this is ready to ship I think.

**Comment by ljwolf** [comment]
test failure looks to be a network error. I'm not sure how to re-run ths?

**Review by martinfleis — APPROVED** [review]
Pushed some minor fixes of docs.
